database/apikeys.go

```diff
@@ -0,0 +1,100 @@
+package database
+
+import (
+	"context"
+	"encoding/base64"
+	"time"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/NebulousLabs/fastrand"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+/**
+API keys are authentication tokens generated by users. They do not expire, thus
+allowing users to use them for a long time and to embed them in apps and on
+machines. API keys can be revoked when they are no longer needed or if they get
+compromised. This is done by deleting them from this service.
+*/
+
+type (
+	// APIKey is a non-expiring authentication token generated on user demand.
+	APIKey struct {
+		ID     primitive.ObjectID `bson:"_id,omitempty" json:"id"`
+		UserID primitive.ObjectID `bson:"user_id" json:"-"`
+		// Key is a base64URL-encoded []byte with length PubKeySize
+		Key       string    `bson:"key" json:"key"`
+		CreatedAt time.Time `bson:"created_at" json:"createdAt"`
+	}
+)
+
+// APIKeyCreate creates a new API key.
+func (db *DB) APIKeyCreate(ctx context.Context, user User) (*APIKey, error) {
+	if user.ID.IsZero() {
+		return nil, errors.New("invalid user")
+	}
+	ak := APIKey{
+		UserID:    user.ID,
+		Key:       base64.URLEncoding.EncodeToString(fastrand.Bytes(PubKeySize)),
+		CreatedAt: time.Now().UTC(),
+	}
+	ior, err := db.staticAPIKeys.InsertOne(ctx, ak)
+	if err != nil {
+		return nil, err
+	}
+	ak.ID = ior.InsertedID.(primitive.ObjectID)
+	return &ak, nil
+}
+
+// APIKeyFetch fetches an API key from the DB.
+func (db *DB) APIKeyFetch(ctx context.Context, user User, id primitive.ObjectID) (*APIKey, error) {
+	if user.ID.IsZero() {
+		return nil, errors.New("invalid user")
+	}
+	var ak APIKey
+	sr := db.staticAPIKeys.FindOne(ctx, bson.M{"user_id": user.ID, "_id": id})
+	if sr.Err() != nil {
+		return nil, sr.Err()
+	}
+	err := sr.Decode(&ak)
+	if err != nil {
+		return nil, err
+	}
+	return &ak, nil
+}
+
+// APIKeyDelete deletes an API key.
+func (db *DB) APIKeyDelete(ctx context.Context, user User, ak string) error {
+	if user.ID.IsZero() {
+		return errors.New("invalid user")
+	}
+	filter := bson.M{
+		"key":     ak,
+		"user_id": user.ID,
+	}
+	_, err := db.staticAPIKeys.DeleteOne(ctx, filter)
+	return err
+}
+
+// APIKeyList lists all API keys that belong to the user.
+func (db *DB) APIKeyList(ctx context.Context, user User) ([]*APIKey, error) {
+	if user.ID.IsZero() {
+		return nil, errors.New("invalid user")
+	}
+	aks := make([]*APIKey, 0)
+	c, err := db.staticAPIKeys.Find(ctx, bson.M{"user_id": user.ID})
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if errDef := c.Close(ctx); errDef != nil {
```

Comment:
I believe this is redundant, you should only close a cursor manually if you don't exhaust it fully, which `.All` does.

Reply:
You are right. I hadn't noticed that All closes the cursor.

Reply:
I think you are right that the cursor gets closed when there are no more elements. But I would still like to keep it here. For two reasons.

1. you never know how the code below changes and always closing a `io.Closer` once you are done with it is just great practice. 
2. I don't know what happens if `c.All` fails halfway through.

Reply:
I disagree but I don't mind. If `c.All` fails the cursor will get closed anyway. This `defer` and extra close is just completely redundant and adds complexity, it's good practice to close your resources obviously but in this particular case it is not necessary, it's very much like closing the request body in the HTTP handlers. But I don't mind if you keep it in. Just for this one time, I disagree with Chris 😛 let it be known.